### ISSUE TITLE
Ninja does not correctly resolve symlinks in depfiles

### ISSUE THREAD

**Issue by kalekundert** [ninja-build/ninja#1400]
I've been having trouble trying to build a large project with ninja, because it keeps trying to rebuild the whole project from scratch every time.  I think I tracked this down to a symlink being resolved incorrectly in the depfile generated by `clang++` (`g++` doesn't have the same problem), but I'll try to explain the bug below:

First, I should say that I'm using ninja 1.8.2 installed from the Fedora repositories.

I started by running `ninja path/to/file.o` twice.  Both runs successfully compiled the file, but the second run shouldn't have needed to.  Next I ran `ninja -d explain path/to/file.o` to see why ninja thought it had to rebuild the file.  I got about 100 dirty files, all of which were stdlib headers.  Here are the first few, for example:

```
ninja explain: output /include/c++/7/cstddef of phony edge with no inputs doesn't exist
ninja explain: /include/c++/7/cstddef is dirty
ninja explain: output /include/c++/7/x86_64-redhat-linux/bits/c++config.h of phony edge with no inputs doesn't exist
ninja explain: /include/c++/7/x86_64-redhat-linux/bits/c++config.h is dirty
ninja explain: output /include/c++/7/x86_64-redhat-linux/bits/os_defines.h of phony edge with no inputs doesn't exist
ninja explain: /include/c++/7/x86_64-redhat-linux/bits/os_defines.h is dirty
ninja explain: output /include/c++/7/x86_64-redhat-linux/bits/cpu_defines.h of phony edge with no inputs doesn't exist
ninja explain: /include/c++/7/x86_64-redhat-linux/bits/cpu_defines.h is dirty
```

Ninja is correct that `/include` does not exist on my machine.  These header dependencies came from a depfile generated by `clang++`, so I ran `ninja -d keepdepfile path/to/file.o` and looked at the resulting depfile.  Here's the `cstddef` header path generated by `clang++`:

```
/bin/../lib/gcc/x86_64-redhat-linux/7/../../../../include/c++/7/cstddef
```

If you resolve this path lexically (i.e. without checking for symlinks), as ninja seems to be doing, you end up with `/include/c++/7/cstddef`, which doesn't exist.  The problem is that on my machine, `/lib` is a symlink to `/usr/lib`.  When that's taken into account, the path generated by `clang++` is resolved to a file that does exist:

```
$ realpath /bin/../lib/gcc/x86_64-redhat-linux/7/../../../../include/c++/7/cstddef
/usr/include/c++/7/cstddef
```

TLDR: It seems like ninja is resolving paths in depfiles without checking for symlinks.  This seems like a pretty easy issue, so I'll poke around and see if I can make a PR to fix this.

**Comment by kalekundert** [comment]
I just realized that this is exactly the same issue as #1330, and is (reasonably) regarded as a bug in either `cmake` or `clang`.

**Comment by kalekundert** [comment]
In case anyone else has the same problem, I worked around it with `sudo ln -s /usr/include/ include`.  I could've also just used `g++`, but I like `clang++` more.

**Comment by evmar** [comment]
Great debugging, glad it worked out for you.  For what it's worth, if we had to resolve paths non-lexically it would probably impact performance a lot (lots of disk accesses to check all the directories).
